internal/kibana/role.go

```diff
@@ -0,0 +1,501 @@
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: apikeySchema,
+	}
+}
+
+func resourceRoleUpsert(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClientFromSDKResource(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	kibana, err := client.GetKibanaClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	queryParams := ""
+	if d.IsNewResource() {
+		queryParams = "?createOnly=true"
+	}
+	kibanaRole := kbapi.KibanaRole{
+		Name:          fmt.Sprintf("%s%s", d.Get("name").(string), queryParams),
+		Kibana:        []kbapi.KibanaRoleKibana{},
+		Elasticsearch: &kbapi.KibanaRoleElasticsearch{},
+	}
+
+	if v, ok := d.GetOk("kibana"); ok {
+		kibanaRole.Kibana, diags = expandKibana(v)
+		if diags != nil {
+			return diags
+		}
+	}
+
+	if v, ok := d.GetOk("elasticsearch"); ok {
+		kibanaRole.Elasticsearch = expandElasticsearch(v)
+	}
+
+	if v, ok := d.GetOk("metadata"); ok {
+		kibanaRole.Metadata, diags = expandMetadata(v)
+		if diags != nil {
+			return diags
+		}
+	}
+
+	roleManageResponse, err := kibana.KibanaRoleManagement.CreateOrUpdate(&kibanaRole)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(roleManageResponse.Name)
+	return resourceRoleRead(ctx, d, meta)
+}
+
+func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClientFromSDKResource(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+
+	name := d.Id()
+
+	kibana, err := client.GetKibanaClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	role, err := kibana.KibanaRoleManagement.Get(name)
+	if role == nil && err == nil {
+		d.SetId("")
+		return diags
+	}
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	// set the fields
+	if err := d.Set("name", role.Name); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("elasticsearch", flattenElasticsearchData(role.Elasticsearch)); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("kibana", flattenKibanaData(&role.Kibana)); err != nil {
+		return diag.FromErr(err)
+	}
+	if role.Metadata != nil {
+		metadata, err := json.Marshal(role.Metadata)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		if err := d.Set("metadata", string(metadata)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	return diags
+}
+
+func resourceRoleDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, diags := clients.NewApiClientFromSDKResource(d, meta)
+	if diags.HasError() {
+		return diags
+	}
+	resourceId := d.Id()
+
+	kibana, err := client.GetKibanaClient()
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = kibana.KibanaRoleManagement.Delete(resourceId)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId("")
+	return diags
+}
+
+// Helper functions
+
+func expandMetadata(v interface{}) (map[string]interface{}, diag.Diagnostics) {
```

Comment:
These will be visible through the entire `internal/kibana` package. It'd be worthwhile to have some more specific naming here. 

```suggestion
func expandKibanaRoleMetadata(v interface{}) (map[string]interface{}, diag.Diagnostics) {
```